### ISSUE TITLE
Self-Service asset checkout model

### ISSUE THREAD

**Issue by snipe** [snipe/snipe-it#5994]
It's a use-case we're starting to see more of, where they want their own users to be able to self-checkout from a list of assets. (This could also work for hackerspaces, etc.)

So basically, allowing unprivileged users to be able to pick their own assets, check out the assets to themselves, add their own note on the checkout, etc. 

I imagine this would likely still need to be company scoped, but I don't see this as being a big deal to do, in fact I've already started it. 

Right now, the sticking point for me is how to implement this from an information architecture/UX perspective. I feel like this might be better served as an option on the users themselves (like activated, etc), versus a user permission or group permission. 

I'm still kicking it around though.

**Comment by EarlRamirez** [comment]
@snipe this looks rather interesting, will there be an option where the admin can mark/tick if this asset can be self checked out? A member of our management team wanted a functionality such as cart where users can add multiple items to a cart, check it out and have it delivered to the user(s). 

In my environment I will like to have this option available based on models because not all users use the same categories.

**Comment by snipe** [comment]
> will there be an option where the admin can mark/tick if this asset can be self checked out

That's literally what I'm asking people here 😁

**Comment by EarlRamirez** [comment]
hahaha, in that case, it is a reasonable scenario and it will definitely be good to manage this from a model level compared to a category or individual asset, at least in the environment that I am in.

**Comment by redlukas** [comment]
I would love that feature! How about you create a new metastatus "Self Checkout" and a user permission "can self-checkout". Then, if i give a user only the "can self-checkout" permission, he should be able to see all assets with the metastatus "Self Checkout" and be able to check them out, but only to himself. 

Nice to have: A user permission "Kiosk mode" that overrides all other permissions. If a User with the "Kiosk mode" permission enabled is logged in i want to be able to (exclusively and in that order)

1. select a user with the "can self-checkout" permission (preferably via barcode/QR code, fallback via dropdown-search),

2. select an asset with the metastatus "Self checkout" (again, preferably by barcode) and 

3. if i have not checked a box "Checkout multiple Items" prior to selecting the item, immediately logout the selected user, to be ready for the next one.

that way i clould walk an entire class by the closet with the laptops, every kid grabs one, scans their user-card, scans their laptop and off they go. This - of course - would mean being able to be able to print labels with barcodes for users, not just assets.

**Comment by snipe** [comment]
@redlukas - So, kiosk mode is actually why we're building this feature. We'll likely provide a plain template without navigation so that it can be used in kiosks in corporations and schools, and those organizations could potentially iframe the kiosk mode screen into their own kiosk chrome . 

I don't think using the status/meta status label here makes sense though. The status label really describes the physical state of the hardware, not the business rules around who has access to it. 

@EarlRamirez I'm thinking asset model makes me the most sense, although what happens when you have 300 MacBook Pros, but only half of them are meant for self-checkout?

**Comment by redlukas** [comment]
@snipe i understand that my proposal includes a use of the metastatus that was not previously intended.  but i do ask myself what the alternatives are. because if you add a checkbox "Can be self-checked-out" similar to the "This Item can be requested" one, the Status of the asset would stay in "ready to deploy", indicating to the admin that they can do with the asset as they please (ie give it out as a long-term loaner, use it as a replacement for a failed item...). But at least in my organisation we have a reservation system for the self-loan items i would like to use the self-checkout for. that would mean i'd have to check for every item that is "Ready to deploy" if i have not put it in the user-accessible closet. To make a long explanation short: By looking at the list, it would be near-impossible to distinguish items at the free disposal of the admin and items potentially promised to a user.

that being said, you have much more insight into what the status/metastatus is intended to be and what the plans are for it, so if you say this does not work/is too much of a hack, ill take your word for it.

**Comment by memtech3** [comment]
Unless it is too hard I think you should allow the admin to require people to type an employee id in (or scan with an HID barcode reader) so that employees can't check items out to other people

**Comment by rkantos** [comment]
For the barcodes - just add QuaggaJS support for the checkout page and be done with it! QuaggaJS would also work wonders elsewhere where barcodes are utilized! The only downside is that it doesn't work with 2D-codes. The demo should already give a good idea on how it could work! https://serratus.github.io/quaggaJS/examples/live_w_locator.html

**Comment by Cabe** [comment]
> Unless it is too hard I think you should allow the admin to require people to type an employee id in (or scan with an HID barcode reader) so that employees can't check items out to other people

That or it defaults it the current logged in user (which is how I would love it to work).

User Flow: 

- User scans item barcode to check it out
- Presents a login page
- Pre-selects current user
- Presents expected check-in dialog (or indefinite)
- Present checkout button. 

This is especially useful on mobile browsers as the dropdown/search is a little fiddly.

**Comment by izharyuk** [comment]
Why not just make Requestable assets available for self-checkout?

**Comment by ukrainets** [comment]
Looks like this feature still in development, just want to assure you it’s a very useful one. 
At my company, we are looking for an inventory/checkout solution for our test devices.
Currently, we are testing Snipe-it which might be useful to some degree but self-checkout would be a key feature for our use as well as many other IT companies who have multiple test devices. Search for "QA device lab" and you will see a big field for potential Snipe-it customers. 
@snipe Are you still planning to implement Self-checkout?

**Comment by toaston** [comment]
I’d just like to throw our hat into the ring for this potential feature.

We’re an educator specialising in sound engineering and music production where students book out equipment to use in our recording studios.  It would not be unreasonable for a student to book out 50+ assets at a time (i.e. 16x mic stands, 16x mics, 16x mic cables, 1x drum kit, 2x amplifiers, 2x guitars, 4x headphones, 2x instrument cables)

We currently use a paper system but have wanted to move to an electronic system for years.  We discovered and installed snipe-it a couple of weeks ago.  There is so much we love about it but haven’t yet found a workflow that doesn’t create significantly more work for our staff.

The ability for students to check out equipment would solve that.  We envisioned a couple of terminals in reception where students would log in with their own LDAP credentials, add assets to a cart which they could review and then confirm by checking-out.  Ideally they could quickly add multiple quantities of the same asset model to cart (i.e mic stands); perhaps similar to the Requestable Assets > Asset Models tab.
Staff would then be able to view the students order via a mobile device, email or printout to gather the assets and hand them over the counter.

Staff would then be able to check in all assets from a student by clicking one button, setting them all to ‘ready to deploy’.   Perhaps checkboxes for each asset at the check in stage to exclude them from the single button check-in feature.  These assets can then be set to un-deployable etc. if they are returned with a fault.

In relation to your question about some assets not being able to be checked out by unprivileged users. Perhaps categories and individual assets would work.  Categories would allow us to keep digital records of assets that aren’t for general use.  Individual assets would allow us to keep a reserve quantity of items for tutor/lesson use.

@snipe – it’d be great to hear if this is a feature you’re still developing.  Happy to provide any further feedback.

**Comment by effridgedurk** [comment]
Id also like to express how great this feature would be.  We just installed Snipe-It for inventory but also checking in/out.  Having a simple, easy process would be fantastic.

**Comment by szabolcs-nagy-4ig** [comment]
We are working with cca. 30 developers / testers and 150 mobile devices. Self checkin and checkout would be a very useful feature, especially from a simplified mobile friendly page, where a user could quickly do it via scanning the barcode then select the self-understanding option.

**Comment by juangarcessonos** [comment]
just expressing my interest in this feature. I have some kiosks i would love to plug into Snipe-IT into

**Comment by pbrink231** [comment]
I see two scenarios with this.

1) Self checkout and grab exact items at that moment
2) user or admin create a checkout like a cart which is then filled and picked up

Each of these are slightly different when it comes to assets.  accessories are about the same for either one.

1) Self checkout of exact items
The user would select the exact assets being chosen.  This could be by scanning the barcode of the item while at the kiosk or searching and choosing.  This is not one of our use cases at the moment but I can imagine a couple scenarios it would be useful for.  For accessories and others would be easy to select and select a quantity

For 2)
This is more the use case we have.  You would make a cart of asset models, consumables and accessories (could add licenses maybe, components dont work well here unless you add then to an asset in your cart).  Can have a quantity selector to easily grab multiple of them.  This would then be filled when it was put in.  In case of asset models, correct assets would be selected.  Accessories/consumables would make sure quantity is available.  This would be where future improvements would be creating a purchase order for missing items.

I can see having 2 work with 1 by being able to fill at the same time.  You could also have them make a checkout and then when they get their they fill it themselves like an admin would.

making asset models allowed to be self ordered vs admin who can order any items.

**Comment by AliBabba3** [comment]
Just wondering if this is still on a future road map?  I would love to have/use this feature

**Comment by viclou** [comment]
+1 FD18832

**Comment by ForrestHerman-Verista** [comment]
+1
I'm currently working on a project where this Kiosk Mode would be absolutely perfect for my needs.

**Comment by izharyuk** [comment]
Any progress on the Self-Service asset checkout?

**Comment by bebeto-web** [comment]
Hi There,
This thread still active ?, I would like to seek your help about the SnipeIT, which I can right a code and manipulate the trigger of the Action buttons (Checkin, Checkout, Update and Delete) to send a data in a 3rd party software system (send in our own database ) once the API was executed after button is triggered. BTW I'm a newbie in SnipeIT sorry for the inconvenience

**Comment by Norjek** [comment]
> 
> 
> Self checkin and checkout would be a very useful feature, especially from a simplified mobile friendly page, where a user could quickly do it via scanning the barcode then select the self-understanding option.

+1 to that

**Comment by bzeus** [comment]
+1 FD21432

**Comment by XeTrinity** [comment]
Any updates on this? I would love a kiosk like system, It would really help employees be able to get items they need from our inventory without needing to ask every time.

**Comment by AHH-BEES** [comment]
+1 to a kiosk type system for general users. I would love this for things like lab laptops

**Comment by adagioajanes** [comment]
I think the first version of this could be similar to something like the bulk audit screen. Either for check in or check out.

On this screen, I would scan an asset, and it would immediately return the asset status. At that point, I could check it in or check it out to a user.

I could also set a 'Default Location' for this kiosk, and put it in check in mode. Every item I scan and press enter (something most barcode scanners can be configured to do, it would automatically check the item in. The scan could either be an asset tag number, or the full url for the asset. That way the person scanning the assets doesn't have to worry about which barcode they scanned.

**Comment by jabums** [comment]
+1
This would be very useful at our school!

**Comment by szabolcs-nagy-4ig** [comment]
Indeed, this feature request has a long history. Allison, can you tell,us how do you see the future of this idea ?
Marry Xmas to all,
Szabi

Get Outlook for iOS<https://ddei5-0-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2faka.ms%2fo0ukef&umid=0E2C23C9-D38F-1105-B8D9-DCC30F2F4702&auth=0bc625e4f964a7f3d8621dba6af342f4b78c0ff1-22cb0bbe4e5288da5130508b9fa4e95ac4167616>
________________________________

**Comment by yokhoe** [comment]
lady @snipe could you update us whether self-service is still a model that is being considered?  TIA

**Comment by snipe** [comment]
Still very much so, yes.

**Comment by Magzter** [comment]
+1

Looking at rolling out this system for a client but self-service checkout/checkin is a must. 
I think it's reasonable to mark records as self-serviceable.

@snipe Is this already in progress or looking for devs to start? If it is where can I find the branch?

**Comment by nghia-dang** [comment]
- It would be useful we could set individual items, a batch of items and all items of type model to be self-checkout capable
- It would be really handy if LDAP groups were supported so that items could be self-checkout available but the items would only be available for self checkout to the one or more LDAP groups specified.

**Comment by zacharyfleck** [comment]
Noticed there's a permission for self checkout with helper text that states "This user may check out assets that are marked for self-checkout." however can't seem to find any way to mark any assets for self checkout. I assume this is still in the works? If so, I hope it'll go beyond assets and allow self-checkout for accessories and consumables too, and hope to see it soon. 🙂

**Comment by Muskelkatermann** [comment]
Was thinking about similar functionality last night. My approuch will be like this:

Self checkout Assets are identified using the location like Headquarter -> Selfcheckout Counter

Selfcheckout counter will have a fixed tablet with barcode / qr Scanner attached.
The tablet presents a simple webinterface to first authenticate the user. The interface would list all currently checked out assets of this user that are qualified for selfcheckout. And two buttons "checkin" and "checkout", which will set the keyboardfocus die hidden textfields that are later triggered by the Scanner Input.

This should be relatively simple using the Snipe-API.

**Comment by omegix** [comment]
+1 for this.  Very interested for our makerspace

**Comment by everR3d** [comment]
Would be awesome to see this. Is there any ETA @snipe ?

**Comment by nghia-dang** [comment]
It would be really handy if permissions could be both individually assigned or assigned by imported LDAP groups (would make life so much easier without a ton of unnecessary management overhead). 

Otherwise, the next best option might be via individuals or groups within the snipeit system, so long as groups can be auto populated externally using the API.

There are many situations where we have multiple groups of people at nested levels, but only want loanable items to be visible to a specific subset for a multitude of reasons..eg funding sources, politics, jealousy, etc.

**Comment by Erzischerzi** [comment]
+1
I'm currently working on a project for my university and it would be absolutely perfect for my needs.

**Comment by jcracine-cimeq** [comment]
+1
See [this related issue](https://github.com/snipe/snipe-it/issues/13825)

**Comment by ndizzlecfizzle** [comment]
+1
looking for this exact feature

**Comment by skysanctuary** [comment]
+1
Would be very helpful

**Comment by biasedlogic** [comment]
Okay, I read this thread but am myself a bit puzzled. There's a permission for 'self checkout' in the group setup, but there's no way to set assets to be self-check-outable :-) or am I missing something?

I'm looking for a setup, where students can check out robot kits on their own, but need a supervisor to check them back in. They have their Snipe-IT accounts, but right now they either can't check out anything themselves, or can check out anything to anyone (not self), so it's kinda broken...

**Comment by jcracine-cimeq** [comment]
> Okay, I read this thread but am myself a bit puzzled. There's a permission for 'self checkout' in the group setup, but there's no way to set assets to be self-check-outable :-) or am I missing something?
> 
> I'm looking for a setup, where students can check out robot kits on their own, but need a supervisor to check them back in. They have their Snipe-IT accounts, but right now they either can't check out anything themselves, or can check out anything to anyone (not self), so it's kinda broken...

You are right. AFAIK it does nothing at the moment.

**Comment by snipe** [comment]
That's correct - we're working on it for v7, but it does have some complexities. We added the permission a while back but are still working through how this should work in a way that best meets everyone's needs, since we have several different use-cases here.

**Comment by jcracine-cimeq** [comment]
> That's correct - we're working on it for v7, but it does have some complexities. We added the permission a while back but are still working through how this should work in a way that best meets everyone's needs, since we have several different use-cases here.

Do you have an idea of an ETA for v7 or it's still in development?

**Comment by hillipp** [comment]
+1 
This would be a greatly appreciated feature.

**Comment by greg0r** [comment]
Unfortunately the feature of self-checkout didn't make it into v7 or am i missing something? Is this feature still on the short list or are we talking more about a longterm feature? Thank you very much to everyone who participates in implementing this feature.

**Comment by Akhun-Delar** [comment]
+1 really needed feature

**Comment by GeoGherkin** [comment]
+1  this feature would be highly appreciated

**Comment by mrdeadlocked** [comment]
+1 for this.  The workflow for us would be that a user would walk up to the kiosk, scan the asset, scan their badge/id, and hit checkout. We could then have a screen to 'checkin' assets at end of the day.

**Comment by snipe** [comment]
@mrdeadlocked The authentication has always been the challenge. *How* are folks logging in via badge, and would every different org actually have badges? What do orgs use if their users don't have badges? Long passwords not stored on a kiosk terminal isn't really feasible.

**Comment by mrdeadlocked** [comment]
In our world it would be easiest to have a checkin/checkout service account in SnipeIT. All checkouts/checkins would be done by that user and whatever permissions/company scope that user has. The end users going to the kiosk would just walk up and be self-assigning their equipment out to themselves. I understand that it would be a more of an 'honor' system on who has what during a given day because of the ability to scan off as someone else if you had their barcode/info.

We give out say 100 of an assets to 'production'. That 100 assets would just be able to 'mostly' be tracked with a self-service process by the end users checkin/checking out daily. The overall 100 assets would always be known to be given to 'production' and not in IT anymore.

**Comment by yurikleb** [comment]
> Okay, I read this thread but am myself a bit puzzled. There's a permission for 'self checkout' in the group setup, but there's no way to set assets to be self-check-outable :-) or am I missing something?
> 
> I'm looking for a setup, where students can check out robot kits on their own, but need a supervisor to check them back in. They have their Snipe-IT accounts, but right now they either can't check out anything themselves, or can check out anything to anyone (not self), so it's kinda broken...

I'm looking to implement a similar functionality - a simple check in/out experience for our lab members.
If I understand correct this is what the "Self Check Out Assets" permission under the "User Self Accounts " tab is for?

However users don't see a "checkout" button unless I also grant them the "View Assets", "Check In" and the "Check Out" permissions under the "Assets" tab.

But once those permissions are granted it the user gets a "Full" Check in/out experience - e.g. they are able to assign an asset to another user, see all item details etc.. This basically also makes the "View Purchase Cost" permission redundant.

I'm on version version v8.3.7 - build 20803 (master) deployed on Azure.

Is the "Self Check Out Assets" permission is still just a "dummy" option?
Whats is the best alternative at the moment? To create a custom kiosk webpage using the API?

**Comment by biasedlogic** [comment]
> Is the "Self Check Out Assets" permission is still just a "dummy" option? Whats is the best alternative at the moment? To create a custom kiosk webpage using the API?

Last time I checked it was still a dummy and the whole snipe-it experience seems to be kinda take-it-or-leave-it, as a lot of the apparent flexibility isn't actually all there. In the end I gave up on it and killed my install a month ago. The self-checkout option hasn't moved since 2024 and without it the tool didn't simplify my life.

**Comment by snipe** [comment]
> and the whole snipe-it experience seems to be kinda take-it-or-leave-it

@biasedlogic thanks for that super-helpful feedback. 🙄 Never mind all of the other work that was done since 2024, of course.

**Comment by yurikleb** [comment]
I'm building a self-checkout Kiosk for our lab using on OAuth. Happy to share it once its ready in the next month or so.
